File: subsystems/knowledge/src/knowledge-graph.ts
/**
 * ==========================================================
 * SKOS
 * Smaily Knowledge Operating System
 * ==========================================================
 *
 * Subsystem : Knowledge
 * Module    : Knowledge Graph
 *
 * Build     : BUILD-000106
 * Sprint    : Sprint 15
 * Version   : 0.0.1
 *
 * Status    : Foundation
 * ==========================================================
 */

export interface KnowledgeNode {

    id: string;

    label: string;

    type: string;

}

export interface KnowledgeEdge {

    id: string;

    source: string;

    target: string;

    relation: string;

}

export class KnowledgeGraph {

    private readonly nodes =
        new Map<string, KnowledgeNode>();

    private readonly edges =
        new Map<string, KnowledgeEdge>();

    /**
     * Register node.
     */
    public addNode(

        node: KnowledgeNode

    ): void {

        this.nodes.set(node.id, node);

    }

    /**
     * Register edge.
     */
    public addEdge(

        edge: KnowledgeEdge

    ): void {

        this.edges.set(edge.id, edge);

    }

    /**
     * Get node.
     */
    public getNode(

        id: string

    ): KnowledgeNode | undefined {

        return this.nodes.get(id);

    }

    /**
     * List nodes.
     */
    public getNodes(): KnowledgeNode[] {

        return Array.from(this.nodes.values());

    }

    /**
     * List edges.
     */
    public getEdges(): KnowledgeEdge[] {

        return Array.from(this.edges.values());

    }

}

File: SKOS-Android/app/src/main/kotlin/com/smaily/skos/knowledge/KnowledgeGraph.kt
package com.smaily.skos.knowledge

import java.util.concurrent.ConcurrentHashMap

/**
 * ------------------------------------------------------------------
 * SKOS (Smaily Knowledge Operating System)
 * ------------------------------------------------------------------
 * KnowledgeGraph
 *
 * Central in-memory knowledge graph.
 *
 * Responsibilities:
 * - Store knowledge nodes
 * - Store knowledge relations
 * - Query graph contents
 *
 * Version : 1.0.0
 * ------------------------------------------------------------------
 */
class KnowledgeGraph {

    /**
     * All knowledge nodes.
     */
    private val nodes =
        ConcurrentHashMap<String, KnowledgeNode>()

    /**
     * All knowledge relations.
     */
    private val relations =
        mutableListOf<KnowledgeRelation>()

    /**
     * Adds a node.
     */
    fun addNode(
        node: KnowledgeNode
    ) {
        nodes[node.id] = node
    }

    /**
     * Adds multiple nodes.
     */
    fun addNodes(
        newNodes: Collection<KnowledgeNode>
    ) {
        newNodes.forEach(::addNode)
    }

    /**
     * Adds a relation.
     */
    fun addRelation(
        relation: KnowledgeRelation
    ) {
        relations.add(relation)
    }

    /**
     * Returns a node by identifier.
     */
    fun findNode(
        id: String
    ): KnowledgeNode? =
        nodes[id]

    /**
     * Returns all nodes.
     */
    fun allNodes(): List<KnowledgeNode> =
        nodes.values.toList()

    /**
     * Returns all relations.
     */
    fun allRelations(): List<KnowledgeRelation> =
        relations.toList()

    /**
     * Removes all graph contents.
     */
    fun clear() {
        nodes.clear()
        relations.clear()
    }

    /**
     * Number of nodes.
     */
    fun nodeCount(): Int =
        nodes.size

    /**
     * Number of relations.
     */
    fun relationCount(): Int =
        relations.size
}
